### PR TITLE
New tests and changes to handle YaST's new libstorage

### PR DESCRIPTION
**Pull request by cwh42**

### CODE REVIEW

File: partitioning_guided.pm [hunk 1]
```diff
@@ -0,0 +1,53 @@
+# SUSE's openQA tests
+#
+# Copyright © 2009-2013 Bernhard M. Wiedemann
+# Copyright © 2012-2016 SUSE LLC
```
Comment (okurz) [inline partitioning_guided.pm]: Please only use current year

File: partitioning_guided.pm [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+#
+# Copying and distribution of this file, with or without modification,
+# are permitted in any medium without royalty provided the copyright
+# notice and this notice are preserved.  This file is offered as-is,
+# without any warranty.
+
+# Summary: Use guided partitioning setup to delete an existing windows partition. 
+# Maintainer: cwh
```
Comment (okurz) [inline partitioning_guided.pm]: Can you use an email instead?

File: partitioning_guided.pm [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+
+use warnings;
+use parent qw(y2logsstep);
```
Comment (okurz) [inline partitioning_guided.pm]: Please use single ticks for single value imports

File: partitioning_guided.pm [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+sub run() {
+    my $self             = shift;
+
+    die "Module only supports StorageNG" if !get_var("STORAGE_NG");
+
+    # Enter guided setup
+    send_key "alt-g";
+
+    assert_screen('inst-partition-guided', 10);
```
Comment (okurz) [inline partitioning_guided.pm]: Just `assert_screen` with default timeout should be fine

File: partitioning_guided.pm [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+sub save_logs_and_resume {
+    my $self = shift;
+    $self->get_to_console;
+    $self->save_upload_y2logs();
+    select_console 'installation';
+}
+
+sub run() {
+    my $self             = shift;
```
Comment (okurz) [inline partitioning_guided.pm]: Just delete if you don't use it

File: partitioning_guided.pm [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+# without any warranty.
+
+# Summary: Use guided partitioning setup to delete an existing windows partition. 
+# Maintainer: cwh
+
+
+use warnings;
+use parent qw(y2logsstep);
+use testapi;
+#use utils 'sle_version_at_least';
```
Comment (okurz) [inline partitioning_guided.pm]: Don't add disabled code as nobody but you will understand why this is not active. Better leave it out and keep in your local git repository, either in `git stash` or a temporary "WIP"-commit

File: tests/installation/grub_test.pm [hunk 7]
```diff
@@ -57,6 +57,18 @@ sub run {
         send_key_until_needlematch 'inst-bootmenu-boot-harddisk', 'up';
         send_key 'ret';
     }
+
+    if (get_var("STORAGE_NG") && get_var("ENCRYPT")) {
+        my @tags = ();
+	for ( my $disk = 0; $disk < get_var("NUMDISKS", 1); $disk++ ) {
```
Comment (okurz) [inline tests/installation/grub_test.pm]: tidy (reason of check fail)

File: tests/installation/installation_overview.pm [hunk 8]
```diff
@@ -32,6 +32,21 @@ sub run {
         assert_screen "inst-xen-pattern";
     }
 
+    # preserve it for the video
+    wait_idle 10;
```
Comment (okurz) [inline tests/installation/installation_overview.pm]: delete. We don't allow "wait_idle" anymore. A rebase should provide a check that fails whenever someone uses a wait_idle

File: tests/installation/installation_overview.pm [hunk 9]
```diff
@@ -32,6 +32,21 @@ sub run {
...
+    # In case the proposal does not fit on the screen, there is a scrollbar shown.
+    # Scroll down to see any errors.
+    if (check_screen('inst-scrollbar', 0)) {
+        send_key 'tab';
+        send_key 'end';
+
+        assert_screen "inst-overview-booting";
+
+        # preserve it for the video
+        wait_idle 10;
```
Comment (okurz) [inline tests/installation/installation_overview.pm]: delete

File: tests/installation/partitioning_expert.pm [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+#
+# Copyright © 2017 SUSE LLC
+#
+# Copying and distribution of this file, with or without modification,
+# are permitted in any medium without royalty provided the copyright
+# notice and this notice are preserved.  This file is offered as-is,
+# without any warranty.
+#
+#
+# Summary:
```
Comment (okurz) [inline tests/installation/partitioning_expert.pm]: That's not a very convincing summary ;-)

File: tests/installation/partitioning_expert.pm [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+use testapi;
+
+sub run() {
+    send_key $cmd{expertpartitioner};
+    assert_screen 'expert-partitioner';
+
+    assert_and_click 'hard-disks';
+    assert_and_click 'home';
+
+    send_key 'alt-d'; # Delete
```
Comment (okurz) [inline tests/installation/partitioning_expert.pm]: I guess this might fail because the next keypress might get lost. Better guard it with `wait_screen_change` -> `wait_screen_change { send_key 'alt-d' };`

File: tests/installation/partitioning_expert.pm [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+    send_key 'alt-y'; # Confirm with 'yes'
+    assert_and_click 'hard-disks';
+    save_screenshot;
+
+    send_key $cmd{accept};
+    die "/home still there" if check_screen('home');
```
Comment (okurz) [inline tests/installation/partitioning_expert.pm]: Please don't use check_screen with default timeout. Either surround the previous `send_key` with `wait_screen_change` or use a multi-tag assert_screen plus `match_has_tag`

File: tests/installation/partitioning_lvm.pm [hunk 13]
```diff
@@ -51,11 +51,23 @@ sub run {
...
+    }
+    else {
+        send_key "alt-d";
+    }
 
-    if (check_screen('inst-partition-radio-buttons', 10)) {    # detect whether new (Radio Buttons) YaST behaviour
+    my $numdisks = get_var("NUMDISKS");
+    print "NUMDISKS = $numdisks\n";
+
+    check_screen([qw(inst-partition-radio-buttons inst-partition-guided inst-partitioning-scheme)], 10);
```
Comment (okurz) [inline tests/installation/partitioning_lvm.pm]: Please don't use check_screen with timeout other than 0. Either surround the previous `send_key` with `wait_screen_change` or use a multi-tag assert_screen plus `match_has_tag`

**Comment by okurz** [comment]
Delete dot at end-of-line of commit message subject lines and PR messages

File: tests/installation/installation_overview.pm (hunk 8 above)
Comment (rwx788) [inline tests/installation/installation_overview.pm]: You can use wait_still_screen instead.

File: tests/installation/partitioning_lvm.pm [hunk 14]
```diff
@@ -75,6 +87,48 @@ sub run {
             $collect_logs = 1;
         }
     }
+    elsif (get_var("STORAGE_NG")) {
+        if ($numdisks <= 1) {
+            die "Guided workflow does not skip disk selection when only one disk!"
+                if !match_has_tag('inst-partitioning-scheme');
```
Comment (rwx788) [inline tests/installation/partitioning_lvm.pm]: Matter of taste, but it's more wordy to use "unless" here and get rid of negation

File: tests/installation/partitioning_lvm.pm [hunk 15]
```diff
@@ -75,6 +87,48 @@ sub run {
...
+        }
+        else {
+            die "Guided workflow does skip disk selection when more than one disk! (NUMDISKS=$numdisks)"
+                if !match_has_tag('inst-partition-guided');
```
Comment (rwx788) [inline tests/installation/partitioning_lvm.pm]: Same as above.

File: tests/installation/partitioning_lvm.pm [hunk 16]
```diff
@@ -75,6 +87,48 @@ sub run {
...
+            send_key 'alt-n';
+        }
+        send_key 'alt-e';
+        assert_screen "inst-partitioning-lvm-enabled";
+        if (get_var("ENCRYPT")) {
+            send_key 'alt-a';
+
+            if (!get_var('ENCRYPT_ACTIVATE_EXISTING')) {
+                assert_screen 'inst-encrypt-password-prompt';
+                #send_key 'tab'; # might be needed again later
```
Comment (rwx788) [inline tests/installation/partitioning_lvm.pm]: Probably we can remove this one

File: tests/installation/partitioning_lvm.pm [hunk 17]
```diff
@@ -75,6 +87,48 @@ sub run {
...
+        assert_screen "inst-partitioning-lvm-enabled";
+        if (get_var("ENCRYPT")) {
+            send_key 'alt-a';
+
+            if (!get_var('ENCRYPT_ACTIVATE_EXISTING')) {
+                assert_screen 'inst-encrypt-password-prompt';
+                #send_key 'tab'; # might be needed again later 
+                type_password;
+                send_key 'tab';
+                type_password;
```
Comment (rwx788) [inline tests/installation/partitioning_lvm.pm]: Here is a little tricky part, as what we have seen sometimes that we press keys to fast and for example start typing password before we got into password field. But let's see if we experience this issue.

File: tests/installation/partitioning_lvm.pm (hunk 14 above)
Comment (cwh42) [inline tests/installation/partitioning_lvm.pm]: For my personal taste actually an "if not" is more obvious than a "unless" – but if ther is any convention about this, I will of course change it.

File: tests/installation/partitioning_lvm.pm (hunk 13 above)
Comment (cwh42) [inline tests/installation/partitioning_lvm.pm]: I just extended the existing code, here. So that `check_screen …, 10;` already was there, before.
Maybe it is worth rewriting the old code anyway?

File: tests/installation/partitioning_lvm.pm (hunk 14 above)
Comment (rwx788) [inline tests/installation/partitioning_lvm.pm]: Nope, that's why I say that it's matter of taste, so no change is really required. Only point that negated expressions usually are evaluated a little slower, but probably unless does it anyway underneath.

**Comment by cwh42** [comment]
Integrated your proposed changes

File: tests/installation/partitioning_lvm.pm (hunk 14 above)
Comment (okurz) [inline tests/installation/partitioning_lvm.pm]: just use "unless" :-)

File: tests/installation/partitioning_lvm.pm (hunk 13 above)
Comment (okurz) [inline tests/installation/partitioning_lvm.pm]: yes

File: tests/installation/grub_test.pm [hunk 18]
```diff
@@ -57,6 +57,18 @@ sub run {
         send_key_until_needlematch 'inst-bootmenu-boot-harddisk', 'up';
         send_key 'ret';
     }
+
+    if (get_var("STORAGE_NG") && get_var("ENCRYPT")) {
+        my @tags = ();
+        for ( my $disk = 0; $disk < get_var("NUMDISKS", 1); $disk++ ) {
```
Comment (rwx788) [inline tests/installation/grub_test.pm]: Travis complains about spaces after and before brackets.

File: tests/installation/partitioning_lvm.pm (hunk 14 above)
Comment (cwh42) [inline tests/installation/partitioning_lvm.pm]: Acually, when reading Damian Conway's "Per Best Practices" I leaned, that using negative modifiers like `unless` and `until` should be avoided as much as possible. See #13 in http://www.perlmonks.org/?node_id=989115

Comment (rwx788) [inline tests/installation/partitioning_lvm.pm]: It's all subjective, but I totally agree with subpoints: Use unless and until only with positive statements so you don't create a double negative logic, which would be twice as confusing! Never follow an unless block with an else block.
If it's used in that way that you can read it in plain English, like "die unless" it's more wordy than "die if not", which is totally perlish and is confusing for some people. I'm fine if we are just consistent, so I do not request a change here. And thanks for the link, it's nice article!

**Comment by cwh42** [comment]
The setup of perlcritic here in travis is different than when run locally on my and Rodion's. Also it does not tell the file, where it found the error. So the problem here is unfixable for me.
I wasted half a day of debugging here and will be on vacation from now on. Contact my team if you want this to be merged before week 35 when I am back.

**Comment by okurz** [comment]
On Wednesday, 16 August 2017 08:14:46 CEST Christopher Hofmann wrote:
> The setup of perlcritic here in travis is different than when run locally on
> my and Rodion's. Also it does not tell the file, where it found the error.
> So the problem here is unfixable for me. I wasted half a day of debugging
> here and will be on vacation from now on. Contact my team if you want this
> to be merged before week 35 when I am back.

It seems you ran perltidy locally but it failed to read the os-autoinst based 
.perltidyrc
so make sure you have an existing "os-autoinst" linked into os-autoinst-
distri-opensuse or call `make prepare`

**Comment by rwx788** [comment]
@okurz We did exactly this and it looks like it uses different profile. In the end main point here is that it's not clear about which file travis is complaining about, otherwise it's ok to fix results after pushing the commit.

**Comment by okurz** [comment]
squashed and merged locally, approved by @rwx788
